plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy

```diff
@@ -425,12 +527,90 @@ class WaveClient {
+    //
+    // Dockerfile template adpated from the Spack package manager
+    // https://github.com/spack/spack/blob/develop/share/spack/templates/container/Dockerfile
+    // LICENSE APACHE 2.0
+    protected String spackRecipeToDockerFile(String recipe) {
+
+        def result = """\
+# Builder image
+FROM ${config.spackOpts().builderImage} as builder
+
+RUN mkdir -p /opt/spack-env \\
+&&  (sed -e 's;compilers:;compilers::;' \\
+         -e 's;^ *flags: *{};    flags:\\n      cflags: ${config.spackOpts().cFlags}\\n      cxxflags: ${config.spackOpts().cxxFlags}\\n      fflags: ${config.spackOpts().fFlags};' \\
+         /root/.spack/linux/compilers.yaml) > /opt/spack-env/compilers.yaml \\
+&&  (echo "spack:" \\
+&&   echo "  include: [/opt/spack-env/compilers.yaml]" \\
+&&   echo "  concretizer:" \\
+&&   echo "    unify: true" \\
+&&   echo "  specs: [${recipe}]" \\
+&&   echo "  packages:" \\
+&&   echo "    all:" \\
+&&   echo "      target: [${config.spackOpts().target}]" \\
+&&   echo "  config:" \\
+&&   echo "    install_tree: /opt/software" \\
+&&   echo "  view: /opt/view") > /opt/spack-env/spack.yaml
+
+# Install packages, clean afterwards
+RUN cd /opt/spack-env && spack env activate . && spack install --fail-fast && spack gc -y
+
+# Strip binaries
+RUN find -L /opt/._view/* -type f -exec readlink -f '{}' \\; | \\
+    xargs file -i | \\
+    grep 'charset=binary' | \\
+    grep 'x-executable\\|x-archive\\|x-sharedlib' | \\
+    awk -F: '{print \$1}' | xargs strip -s
+
+RUN cd /opt/spack-env && \
+    spack env activate --sh -d . >> /etc/profile.d/z10_spack_environment.sh && \
+    original_view=\$( cd /opt ; ls -1d ._view/* ) && \
+    sed -i "s;/view/;/\$original_view/;" /etc/profile.d/z10_spack_environment.sh && \
+    echo "# Needed for Perl applications" >>/etc/profile.d/z10_spack_environment.sh && \
+    echo "export PERL5LIB=\$(eval ls -d /opt/._view/*/lib/5.*):\$PERL5LIB" >>/etc/profile.d/z10_spack_environment.sh && \
+    rm -rf /opt/view
+
+# Runner image
+FROM ${config.spackOpts().runnerImage}
+
+COPY --from=builder /opt/spack-env /opt/spack-env
+COPY --from=builder /opt/software /opt/software
+COPY --from=builder /opt/._view /opt/._view
+COPY --from=builder /etc/profile.d/z10_spack_environment.sh /etc/profile.d/z10_spack_environment.sh
+
+# Near OS-agnostic package addition
+RUN ( apt update -y && apt install -y procps ${config.spackOpts().osPackages} && rm -rf /var/lib/apt/lists/* ) || \\
+    ( yum install -y procps ${config.spackOpts().osPackages} && yum clean all && rm -rf /var/cache/yum ) || \\
+    ( zypper ref && zypper install -y procps ${config.spackOpts().osPackages} && zypper clean -a ) || \\
+    ( apk update && apk add --no-cache ${config.spackOpts().osPackages} && rm -rf /var/cache/apk )
+""" //.stripIndent()
+
+        result = addCommands(result)
+
+        result += """\
+
+ENTRYPOINT ["/bin/bash", "--rcfile", "/etc/profile", "-l", "-c", "\$*", "--" ]
+CMD [ "/bin/bash" ]
+"""//.stripIndent()
```

Comment:
Would not be better to use a template file instead of a big inline string?

Reply:
good point, will be back on this one too

Reply:
Have a look a this class 

https://github.com/nextflow-io/nextflow/blob/762ba19c70006c84070bf1c736825e483c7fdde7/modules/nextflow/src/main/groovy/nextflow/executor/BashTemplateEngine.groovy#L31-L45

Reply:
done